### PR TITLE
docs: Updates for Zoe facet parameters

### PR DESCRIPTION
**Pull request by Chris-Hibbert**

### CODE REVIEW

File: main/zoe/api/zoe-helpers.md [hunk 1]
```diff
@@ -147,6 +147,53 @@ canTradeWith(leftInvite, rightInvite)
...
+
+Checks if the `give` and `want` of two invites would satisfy offer
+safety if the two allocations are swapped mapping the keywords as specified.
+
+```js
+import { makeZoeHelpers } from '@agoric/zoe/src/contractSupport/zoeHelpers';
+
+const { canTradeWithMapKeywords } = makeZoeHelpers(zoe);
+
+const leftInvite = harden({
```
Comment (katelynsills) [inline main/zoe/api/zoe-helpers.md:165]: I think this part is confusing. It's called Invite, but it's not an invite, it's the proposal. And the helper itself takes offerHandles - so there's a couple of things going on that would throw the user, I think.

Comment (Chris-Hibbert) [reply main/zoe/api/zoe-helpers.md]: thanks. I made the handles in the example opaque.

Comment (katelynsills) [reply main/zoe/api/zoe-helpers.md]: Looks good! 👍

File: main/zoe/api/records.md [hunk 2]
```diff
@@ -166,3 +166,8 @@ The possible records are:
   currentAllocation, // the amountKeywordRecord that will be turned into payouts
 }
 ```
+
+## BrandKeywordRecord
+
+The keys are keywords and the values are brands.
```
Comment (tyg) [inline main/zoe/api/records.md:172]: This seems a bit sparce...No example?

File: main/zoe/api/zoe-contract-facet.md [hunk 3]
```diff
@@ -4,23 +4,20 @@
 
 A Zoe Contract Facet is an API object for a running contract instance to access the Zoe state for that instance. A Zoe Contract Facet is accessed synchronously from within the contract, and usually is referred to in code as `zcf`. The contract instance is launched by `E(zoe).makeInstance`, and is given access to the `zcf` object during that launch. In the operation below, the `instanceHandle` is the handle for the running contract instance.
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:5]: That's an awfully complicated first sentence. Suggest a change to "A Zoe Contract Facet is an API object for a running contract instance. It accesses the instance's Zoe state." I'd also change "usually is" to "is usually". Change the third sentence to "The contract instance is launched and given access to the `zcf` object by `E(zoe).makeInstance`." Finally, "Below, `instanceHandle` is the running contract instance's handle."

File: main/zoe/api/zoe-contract-facet.md [hunk 4]
```diff
@@ -4,23 +4,20 @@
...
 
-## zcf.reallocate(offerHandles, newAmountKeywordRecords, sparseKeywords)
+## zcf.reallocate(offerHandles, newAmountKeywordRecords)
+
 - `offerHandles` <router-link to="/glossary/#handle">`{Array <Handle>}`</router-link>
 - `newAmountKeywordRecords` <router-link to="/zoe/api/records.html#amountkeywordrecord">`{Array <AmountKeywordRecord>}`</router-link>
-- `sparseKeywords` `{Array <String>}` sparseKeywords is an optional array of string keywords, which may be a subset of allKeywords.
 
 Instruct Zoe to try to reallocate payouts for the given `offerHandles`.  This will only succeed if the reallocation 1) conserves rights, and 2) is 'offer-safe' for all parties involved. This reallocation is partial, meaning that it applies only to
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:12]: I'd tighten this to:
"Instruct Zoe to try to reallocate payouts for the given `offerHandle`s. This succeeds only if the reallocation:
- Conserves rights.
-.Is 'offer-safe' for all involved parties.
This reallocation is partial, so it only applies to the amounts associated with passed in `offerHandle`s. We ensure that with each reallocation, rights are conserved and offer safety enforced for all offers. This is true even though it is a partial reallocation. Once these invariants are true, they stay true until changes happen."

File: main/zoe/api/zoe-contract-facet.md [hunk 5]
```diff
@@ -4,23 +4,20 @@
...
 the amount associated with the offerHandles that are passed in.  We are able to ensure that with each reallocation,
 rights are conserved and offer safety is enforced for all offers, even though the reallocation is partial, because once
 these invariants are true, they will remain true until changes are made.
 
-newAmountKeywordRecords is an array of `AmountKeywordRecords`, which are objects where the keys are keywords and the
-values are the amounts to be paid to the offer at the same index in the `offerHandles`.
+newAmountKeywordRecords is an array of `AmountKeywordRecords`, which are objects where the keys are
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:17]: `newAmountRecords`

File: main/zoe/api/zoe-contract-facet.md [hunk 6]
```diff
@@ -4,23 +4,20 @@
...
+keywords and the values are the amounts to be paid to the offer at the same index in the `offerHandles`.
 
-This operation throws an error:
-- If any of the newAmountKeywordRecords do not have a value for all the keywords in sparseKeywords. 
-- If any newAmountKeywordRecords have keywords not in sparseKeywords.
-- If there are only 0 or 1 offerHandles given.
+This operation throws an error if there are only 0 or 1 offerHandles given.
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:20]: `offerHandle`s

File: main/zoe/api/zoe-contract-facet.md [hunk 7]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
+## zcf.getCurrentAllocation(offerHandle, brandKeywordRecord)
 - `offerHandle` <router-link to="/glossary/#handle">`{Array <Handle>}`</router-link>
-- `sparseKeywords` sparseKeywords is an array of string keywords, which may be a subset of allKeywords.
-- Returns: <router-link to="/zoe/api/records.html#amount-keyword-record">`{<AmountKeywordRecord>}`</router-link>
+- `brandKeywordRecord` <router-link to="/zoe/api/records.html#brand-keyword-record">
+`{[<BrandKeywordRecord>]}`</router-link> is an optional record mapping keywords to brands for each of the offers.
+- Returns: <router-link to="/zoe/api/records.html#amount-keyword-record">
+`{<AmountKeywordRecord>}`</router-link>
+
+Get the amounts currently allocated to the offer for the keywords specified by brandKeywordRecord. If
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:173]: `brandKeywordRecord`

File: main/zoe/api/zoe-contract-facet.md [hunk 8]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
+brandKeywordRecord is not specified, amounts are returned for only the Keywords that currently have
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:174]: `brandKeywordRecord`

File: main/zoe/api/zoe-contract-facet.md [hunk 9]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
+allocated amounts (including empty amounts.) This might mean that amounts that will be included in the
+payout could be omitted if not specified.
 
-Get the amounts associated with the sparseKeywords for the offer.
+If brandKeywordRecord is specified and amounts for some keywords haven't been assigned to the
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:178]: Change to:  'If `brandKeywordRecord` is specified, empty amounts are filled in for keywords without amounts assigned to the offer.'

File: main/zoe/api/zoe-contract-facet.md [hunk 10]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
+## zcf.getCurrentAllocations(offerHandles, brandKeywordRecords)
 - `offerHandles` <router-link to="/glossary/#handle">`{Array <Handle>}`</router-link>
-- `sparseKeywords` sparseKeywords is an array of string keywords, which may be a subset of allKeywords.
+- `brandKeywordRecords` <router-link to="/zoe/api/records.html#brand-keyword-record">
+`{[ Array <BrandKeywordRecords>]}`</router-link> brandKeywordRecords is an optional array of records
+mapping keywords to brands for each of the offers.
 - Returns: <router-link to="/zoe/api/records.html#amount-keyword-record">`{[<AmountKeywordRecord>]}`</router-link>
 
-Get a list of the amounts associated with the sparseKeywords for the offers.
+Get the amounts currently allocated to each offer for the keywords specified by brandKeywordRecord. If
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:192]: OK, so it looks like there's just a lot of them; brandKeywordRecord should be codified `brandKeywordRecord` throughout.

File: main/zoe/api/zoe-contract-facet.md [hunk 11]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
 const { foo, bar } = zoe.getCurrentAllocation(offerHandle, ['foo', 'bar']);
 ```
 
-## zcf.getCurrentAllocations(offerHandles, sparseKeywords)
+## zcf.getCurrentAllocations(offerHandles, brandKeywordRecords)
 - `offerHandles` <router-link to="/glossary/#handle">`{Array <Handle>}`</router-link>
-- `sparseKeywords` sparseKeywords is an array of string keywords, which may be a subset of allKeywords.
+- `brandKeywordRecords` <router-link to="/zoe/api/records.html#brand-keyword-record">
+`{[ Array <BrandKeywordRecords>]}`</router-link> brandKeywordRecords is an optional array of records
+mapping keywords to brands for each of the offers.
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:189]: "each offer"

File: main/zoe/api/zoe-contract-facet.md (hunk 10 above)
Comment (tyg) [reply main/zoe/api/zoe-contract-facet.md:192]: This is complicated, with multiple objects and relationships, and its running backwards from what they're looking for.  How about:
"For each `brandKeywordRecord specified keyword, from each offer get its currently allocated amount."

File: main/zoe/api/zoe-contract-facet.md [hunk 12]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
 - Returns: <router-link to="/zoe/api/records.html#amount-keyword-record">`{[<AmountKeywordRecord>]}`</router-link>
 
-Get a list of the amounts associated with the sparseKeywords for the offers.
+Get the amounts currently allocated to each offer for the keywords specified by brandKeywordRecord. If
+brandKeywordRecord is not specified, amounts are returned for only the Keywords that currently have
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:193]: "amounts are only returned for the Keywords...."

File: main/zoe/api/zoe-contract-facet.md [hunk 13]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
+allocated amounts (including empty amounts.) This might mean that amounts that will be included in the
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:194]: "Amounts included in the payout might be omitted if not specified."

File: main/zoe/api/zoe-contract-facet.md [hunk 14]
```diff
@@ -166,23 +163,39 @@ Zoe.
...
+payout could be omitted if not specified.
+
+If brandKeywordRecord is specified and amounts for some keywords haven't been assigned to the
```
Comment (tyg) [inline main/zoe/api/zoe-contract-facet.md:197]: If `brandKeywordRecord is specified, and some keywords' amounts haven't been assigned to the offer, empty amounts are filled in.

File: main/zoe/api/zoe-helpers.md [hunk 15]
```diff
@@ -147,6 +147,27 @@ canTradeWith(leftInvite, rightInvite)
...
 ```
 
+## zoeHelpers.canTradeWithMapKeywords(leftOfferHandle, rightOfferHandle, keywords)
+- `leftOfferHandle`
+- `rightOfferHandle`
+- `keywords` - correspondence map of keywords on the left and right handles. keywords is an
+  array containing two equal-length arrays.
+- Returns: `{Boolean}`
+
+Checks if the `give` and `want` of two invites would satisfy offer
```
Comment (tyg) [inline main/zoe/api/zoe-helpers.md:157]: This description of the handle arguments should also be up in the spec on lines 151-2.

File: main/zoe/api/zoe-helpers.md [hunk 16]
```diff
@@ -244,16 +221,13 @@ makeEmptyOffer().then(offerHandle => {...})
...
+- `amount` - the amount to be escrowed. This should be equal to the payment amount
 - `payment` - the payment that will be escrowed
-- `keyword` - the keyword under which the payment should be escrowed.
-  This will be used to create the proposal and the
-  paymentKeywordRecord
-- `recipientHandle` - the offerHandle that we will reallocate the
-  amount to
+- `keyword` - the keyword under which the payment should be escrowed. This will be used to create
+  the proposal and the paymentKeywordRecord.
+- `recipientHandle` - the offerHandle that we will reallocate the amount to
```
Comment (tyg) [inline main/zoe/api/zoe-helpers.md:230]: `offerHandle`

File: main/zoe/api/zoe-helpers.md [hunk 17]
```diff
@@ -244,16 +221,13 @@ makeEmptyOffer().then(offerHandle => {...})
...
-  payment amount
+- `amount` - the amount to be escrowed. This should be equal to the payment amount
 - `payment` - the payment that will be escrowed
-- `keyword` - the keyword under which the payment should be escrowed.
-  This will be used to create the proposal and the
-  paymentKeywordRecord
-- `recipientHandle` - the offerHandle that we will reallocate the
-  amount to
+- `keyword` - the keyword under which the payment should be escrowed. This will be used to create
+  the proposal and the paymentKeywordRecord.
```
Comment (tyg) [inline main/zoe/api/zoe-helpers.md:229]: `paymentKeywordRecord`

File: main/zoe/api/zoe.md [hunk 18]
```diff
@@ -183,24 +183,36 @@ const {
...
-## E(zoe).getCurrentAllocation(offerHandle, sparseKeywords)
+## E(zoe).getCurrentAllocation(offerHandle, brandKeywordRecord)
 - `offerHandle` <router-link to="/glossary/#handle">`{Array <Handle>}`</router-link>
-- `sparseKeywords` sparseKeywords is an array of string keywords, which may be a subset of allKeywords.
+- `brandKeywordRecord` <router-link to="/zoe/api/records.html#brand-keyword-record">`{[<BrandKeywordRecord>]}`</router-link>
+brandKeywordRecord is an optional record mapping keywords to brands for each of the offers.
 - Returns: <router-link to="/zoe/api/records.html#offer-record">`{<AmountKeywordRecord>}`</router-link>
 
-Get the amounts associated with the sparseKeywords for the offer.
+Get the amounts associated with the keywords for the offer. If brandKeywordRecord is not
```
Comment (tyg) [inline main/zoe/api/zoe.md:192]: Let me try to clarify the phrasing;

For the offer's keywords, get their associated amounts. If `brandKeywordRecord` is not specified
only keywords with currently allocated amounts (including empty amounts) have their amounts returned.
If amounts to be included in the payout are not specified, they may be omitted.

File: main/zoe/api/zoe.md [hunk 19]
```diff
@@ -183,24 +183,36 @@ const {
...
+specified, amounts are returned for only the Keywords that currently have allocated amounts
+(including empty amounts.) This might mean that amounts that will be included in the payout
+could be omitted if not specified.
+
+If brandKeywordRecord is specified, and amounts for some keywords haven't been assigned to the
```
Comment (tyg) [inline main/zoe/api/zoe.md:197]: If `brandKeywordRecord` is specified, and some keyword amounts aren't assigned to the offer, empty amounts are returned.

File: main/zoe/api/zoe.md [hunk 20]
```diff
@@ -183,24 +183,36 @@ const {
...
+- `brandKeywordRecords` <router-link to="/zoe/api/records.html#brand-keyword-record">`{[ Array <BrandKeywordRecords>]}`</router-link> brandKeywordRecords is an optional array of records mapping keywords to brands for each of the offers.
 - Returns: <router-link to="/zoe/api/records.html#amount-keyword-record">`{ Array AmountKeywordRecord>}`</router-link>
 
-Get the amounts associated with the sparseKeywords for the offers.
+Get the amounts associated with the keywords for the offers. If brandKeywordRecords is not
+specified, amounts are returned for only the Keywords that currently have allocated amounts
+(including empty amounts.) This might mean that amounts that will be included in the payout
+could be omitted if not specified.
 
+If brandKeywordRecords is specified, and amounts for some keywords haven't been assigned for some
```
Comment (tyg) [inline main/zoe/api/zoe.md:214]: If `brandKeywordRecord` is specified, and some keyword amounts aren't assigned to the offer, empty amounts are returned.

File: main/zoe/api/zoe.md [hunk 21]
```diff
@@ -183,24 +183,36 @@ const {
...
 
-## E(zoe).getCurrentAllocations(offerHandles, sparseKeywords)
+## E(zoe).getCurrentAllocations(offerHandles, brandKeywordRecords)
 - `offerHandles` <router-link to="/glossary/#handle">`{Array <Handle>}`</router-link>
-- `sparseKeywords` sparseKeywords is an array of string keywords, which may be a subset of allKeywords.
+- `brandKeywordRecords` <router-link to="/zoe/api/records.html#brand-keyword-record">`{[ Array <BrandKeywordRecords>]}`</router-link> brandKeywordRecords is an optional array of records mapping keywords to brands for each of the offers.
 - Returns: <router-link to="/zoe/api/records.html#amount-keyword-record">`{ Array AmountKeywordRecord>}`</router-link>
 
-Get the amounts associated with the sparseKeywords for the offers.
+Get the amounts associated with the keywords for the offers. If brandKeywordRecords is not
```
Comment (tyg) [inline main/zoe/api/zoe.md:209]: For the offer's keywords, get their associated amounts. If `brandKeywordRecord` is not specified
only keywords with currently allocated amounts (including empty amounts) have their amounts returned.
If amounts to be included in the payout are not specified, they may be omitted.

**Pull request merged**